### PR TITLE
Make ViT a Function Subclass Model

### LINKED ISSUE #1430: Make ViT a Functional subclass model & remove closures
This should follow the example of #1412 and #1401, including

- Docstring updates
- Colab demonstrating that weight loading works (if pre-trained weights exist) and numerics are identical

### PR DESCRIPTION
**Pull request by cosmo3769**
# What does this PR do?

Fixes #1430 

This [Colab](https://colab.research.google.com/drive/14G1q9yQNBX_1R36QApHVAL9fWa4IePbL?usp=sharing) verifies that existing pretrained checkpoints can be loaded.

# Who can review?

@ianstenbit @LukeWood

### CODE REVIEW

File: keras_cv/models/vit.py [hunk 1]
```diff
@@ -244,69 +224,125 @@ def ViT(
...
+        self.project_dim=project_dim
+        self.mlp_dim=mlp_dim
+        self.classifier_activation=classifier_activation
+
+    def get_config(self):
+        return {
+            "include_rescaling": self.include_rescaling,
+            "include_top": self.include_top,
+            "name": self.name,
+            "input_shape": self.input_shape,
```
Comment (ianstenbit) [inline keras_cv/models/vit.py:328]: This should be:

```
# Remove batch dimension from `input_shape`
"input_shape": self.input_shape[1:],
```

File: keras_cv/models/vit.py [hunk 2]
```diff
@@ -244,69 +224,125 @@ def ViT(
...
+            "classes": self.classes,
+            "patch_size": self.patch_size,
+            "transformer_layer_num": self.transformer_layer_num,
+            "num_heads": self.num_heads,
+            "mlp_dropout": self.mlp_dropout,
+            "attention_dropout": self.attention_dropout,
+            "activation": self.activation,
+            "project_dim": self.project_dim,
+            "mlp_dim": self.mlp_dim,
+            "classifier_activation": self.classifier_activation,
```
Comment (ianstenbit) [inline keras_cv/models/vit.py:340]: please also include 
```
"trainable": self.trainable,
```

**Review by ianstenbit — CHANGES_REQUESTED** [review]
Thank you!
Colab looks good.

Just a few small changes and then we can merge!

File: keras_cv/models/vit.py [hunk 3]
```diff
@@ -174,36 +174,16 @@
...
-    num_heads=None,
-    mlp_dropout=None,
-    attention_dropout=None,
-    activation=None,
-    project_dim=None,
-    mlp_dim=None,
-    classifier_activation="softmax",
-    **kwargs,
-):
+class ViT(keras.Model):
```
Comment (ianstenbit) [inline keras_cv/models/vit.py:177]: Please add
```
@keras.utils.register_keras_serializable(package="keras_cv.models")
```

File: keras_cv/models/vit.py (hunk 1 above)
Comment (cosmo3769) [reply keras_cv/models/vit.py]: Added.

File: keras_cv/models/vit.py (hunk 2 above)
Comment (cosmo3769) [reply keras_cv/models/vit.py:341]: Added

File: keras_cv/models/vit.py (hunk 3 above)
Comment (cosmo3769) [reply keras_cv/models/vit.py:178]: Added.

**Comment by cosmo3769** [comment]
Thank you for reviewing. I have made the necessary changes. Please check if it's good to go. @ianstenbit

**Review by ianstenbit — APPROVED** [review]
Thank you -- looks good!

**Comment by ianstenbit** [comment]
@cosmo3769 Please take a look at the code format issue.

**Comment by cosmo3769** [comment]
> @cosmo3769 Please take a look at the code format issue.

Done.

**Pull request merged**